builtin/logical/pki/storage.go

```diff
@@ -0,0 +1,176 @@
+	SerialNumber string   `json:"serial_number" structs:"serial_number" mapstructure:"serial_number"`
+}
+
+func listKeys(ctx context.Context, s logical.Storage) ([]keyId, error) {
+	strList, err := s.List(ctx, keyPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	keyIds := make([]keyId, 0, len(strList))
+	for _, entry := range strList {
+		keyIds = append(keyIds, keyId(entry))
+	}
+
+	return keyIds, nil
+}
+
+func fetchKeyById(ctx context.Context, s logical.Storage, keyId keyId) (*key, error) {
+	keyEntry, err := s.Get(ctx, keyPrefix+keyId.String())
+	if err != nil {
+		return nil, errutil.InternalError{Err: fmt.Sprintf("unable to fetch pki key: %v", err)}
+	}
+	if keyEntry == nil {
+		// FIXME: Dedicated/specific error for this?
+		return nil, errutil.UserError{Err: fmt.Sprintf("pki key id %s does not exist", keyId.String())}
+	}
+
+	var key key
+	if err := keyEntry.DecodeJSON(&key); err != nil {
+		return nil, errutil.InternalError{Err: fmt.Sprintf("unable to decode pki key with id %s: %v", keyId.String(), err)}
+	}
+
+	return &key, nil
+}
+
+func writeKey(ctx context.Context, s logical.Storage, key key) error {
+	keyId := key.ID
+
+	json, err := logical.StorageEntryJSON(keyPrefix+keyId.String(), key)
+	if err != nil {
+		return err
+	}
+
+	return s.Put(ctx, json)
+}
+
+func listIssuers(ctx context.Context, s logical.Storage) ([]issuerId, error) {
+	strList, err := s.List(ctx, issuerPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	issuerIds := make([]issuerId, 0, len(strList))
+	for _, entry := range strList {
+		issuerIds = append(issuerIds, issuerId(entry))
+	}
+
+	return issuerIds, nil
+}
+
+func fetchIssuerById(ctx context.Context, s logical.Storage, issuerId issuerId) (*issuer, error) {
+	issuerEntry, err := s.Get(ctx, issuerPrefix+issuerId.String())
+	if err != nil {
+		return nil, errutil.InternalError{Err: fmt.Sprintf("unable to fetch pki issuer: %v", err)}
+	}
+	if issuerEntry == nil {
+		// FIXME: Dedicated/specific error for this?
+		return nil, errutil.UserError{Err: fmt.Sprintf("pki issuer id %s does not exist", issuerId.String())}
+	}
+
+	var issuer issuer
+	if err := issuerEntry.DecodeJSON(&issuer); err != nil {
+		return nil, errutil.InternalError{Err: fmt.Sprintf("unable to decode pki issuer with id %s: %v", issuerId.String(), err)}
+	}
+
+	return &issuer, nil
+}
+
+func writeIssuer(ctx context.Context, s logical.Storage, issuer issuer) error {
+	issuerId := issuer.ID
+
+	json, err := logical.StorageEntryJSON(issuerPrefix+issuerId.String(), issuer)
+	if err != nil {
+		return err
+	}
+
+	return s.Put(ctx, json)
+}
+
+func setKeysConfig(ctx context.Context, s logical.Storage, config map[string]keyId) error {
+	json, err := logical.StorageEntryJSON(keyConfig, config)
+	if err != nil {
+		return err
+	}
+
+	return s.Put(ctx, json)
+}
+
+func getKeysConfig(ctx context.Context, s logical.Storage) (map[string]keyId, error) {
+	keyConfigEntry, err := s.Get(ctx, keyConfig)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I think you need a `keyConfigEntry == nil` check and return an empty config (no error). This shouldn't happen on migration, but aligns us with the `getConfig` from storage methods (`url` or `crl`).